dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/model/ResourceLink.java

```diff
@@ -96,47 +108,52 @@ public String toString() {
-                        "Can only build Resource Links out of content with type `File Asset`.",getClass())
+                        "Can only build Resource Links out of content with type `File Asset` or `DotAsset`.",getClass())
                 );
             }
 
+            return build(request, user, contentlet, FileAssetAPI.BINARY_FIELD);
+        }
+
+        public final ResourceLink build(final HttpServletRequest request, final User user, final Contentlet contentlet, final String velocityVarName) throws DotDataException, DotSecurityException {
+
+            final File binary           = Try.of(()->contentlet.getBinary(velocityVarName)).getOrNull();
             final Identifier identifier = getIdentifier(contentlet);
-            if (identifier != null && InodeUtils.isSet(identifier.getInode())){
+            if (binary==null || identifier == null && UtilMethods.isEmpty(identifier.getInode())){
 
-                final boolean downloadRestricted = isDownloadPermissionBasedRestricted(contentlet, user);
+                return new ResourceLink(StringPool.BLANK, StringPool.BLANK, StringPool.BLANK, null, StringPool.BLANK, false, true);
+            }
 
-                final StringBuilder resourceLink = new StringBuilder();
-                final StringBuilder resourceLinkUri = new StringBuilder();
+            final boolean downloadRestricted    = isDownloadPermissionBasedRestricted(contentlet, user);
+            final StringBuilder resourceLink    = new StringBuilder();
+            final StringBuilder resourceLinkUri = new StringBuilder();
 
-                Host host = getHost((String)request.getAttribute(HOST_REQUEST_ATTRIBUTE) , user);
-                if(null == host){
-                   host = getHost(contentlet.getHost(), user);
-                }
+            Host host = getHost((String)request.getAttribute(HOST_REQUEST_ATTRIBUTE) , user);
+            if(null == host){
+                host = getHost(contentlet.getHost(), user);
+            }
 
-                if(request.isSecure()){
-                    resourceLink.append(HTTPS_PREFIX);
-                }else{
-                    resourceLink.append(HTTP_PREFIX);
-                }
-                resourceLink.append(host.getHostname());
-                if(request.getServerPort() != 80 && request.getServerPort() != 443){
-                    resourceLink.append(StringPool.COLON).append(request.getServerPort());
-                }
-                resourceLinkUri.append(identifier.getParentPath()).append(contentlet.getStringProperty(FileAssetAPI.FILE_NAME_FIELD));
-                resourceLink.append(UtilMethods.encodeURIComponent(resourceLinkUri.toString()));
-                resourceLinkUri.append(LANG_ID_PARAM).append(contentlet.getLanguageId());
-                resourceLink.append(LANG_ID_PARAM).append(contentlet.getLanguageId());
+            resourceLink.append(request.isSecure()? HTTPS_PREFIX:HTTP_PREFIX);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [StringBuffer (or StringBuilder).append is called consecutively without reusing the target variable.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5372524)